Source/Diagnostics/ReducedDiags/ParticleHistogram2D.cpp

```diff
@@ -0,0 +1,340 @@
+    Real const bin_min_ord  = m_bin_min_ord;
+    Real const bin_size_ord = m_bin_size_ord;
+    const bool is_unity_particle_weight =
+            (m_norm == NormalizationType::unity_particle_weight) ? true : false;
+
+    bool const do_parser_filter = m_do_parser_filter;
+
+    int const nlevs = std::max(0, myspc.finestLevel()+1);
+    for (int lev = 0; lev < nlevs; ++lev) {
+#ifdef AMREX_USE_OMP
+#pragma omp parallel if (amrex::Gpu::notInLaunchRegion())
+#endif
+        {
+            for (WarpXParIter pti(myspc, lev); pti.isValid(); ++pti)
+            {
+                auto const GetPosition = GetParticlePosition(pti);
+
+                auto & attribs = pti.GetAttribs();
+                ParticleReal* const AMREX_RESTRICT d_w = attribs[PIdx::w].dataPtr();
+                ParticleReal* const AMREX_RESTRICT d_ux = attribs[PIdx::ux].dataPtr();
+                ParticleReal* const AMREX_RESTRICT d_uy = attribs[PIdx::uy].dataPtr();
+                ParticleReal* const AMREX_RESTRICT d_uz = attribs[PIdx::uz].dataPtr();
+
+                long const np = pti.numParticles();
+
+                //Flag particles that need to be copied if they cross the z_slice
+                amrex::ParallelFor(np,
+                                   [=] AMREX_GPU_DEVICE(int i)
+                                   {
+                                       amrex::ParticleReal x, y, z;
+                                       GetPosition(i, x, y, z);
+                                       auto const w  = (amrex::Real)d_w[i];
+                                       auto const ux = d_ux[i] / PhysConst::c;
+                                       auto const uy = d_uy[i] / PhysConst::c;
+                                       auto const uz = d_uz[i] / PhysConst::c;
+
+                                       // don't count a particle if it is filtered out
+                                       if (do_parser_filter)
+                                           if (!fun_filterparser(t, x, y, z, ux, uy, uz))
+                                               return;
+
+                                       // continue function if particle is not filtered out
+                                       auto const f_abs = fun_partparser_abs(t, x, y, z, ux, uy, uz);
+                                       auto const f_ord = fun_partparser_ord(t, x, y, z, ux, uy, uz);
+
+                                       // determine particle bin
+                                       int const bin_abs = int(Math::floor((f_abs-bin_min_abs)/bin_size_abs));
+                                       if ( bin_abs<0 || bin_abs>=num_bins_abs ) return; // discard if out-of-range
+
+                                       int const bin_ord = int(Math::floor((f_ord-bin_min_ord)/bin_size_ord));
+                                       if ( bin_ord<0 || bin_ord>=num_bins_ord ) return; // discard if out-of-range
+
+                                       amrex::Real &data = d_table(bin_abs, bin_ord);
```

Comment:
Is this line causing the CUDA compile error?
I thought `Table2D` can be used on device, but maybe it's another line that causes this.